db/api/v1/db/projects.db.js

```diff
@@ -0,0 +1,162 @@
+/**
+ * Projects table database interaction layer. These functions are exported to
+ * provide the ability to query Project specific data.
+ * 
+ * TODO: Insert License
+ */
+
+const utils = require('../utils');
+
+// A query to select projects from the database
+const QUERY_FIND = `
+    SELECT 
+        id,
+        sponsor_id, 
+        name, 
+        description
+    FROM 
+        project 
+    WHERE 
+        TRUE 
+`;
+
+// A query to create a new project in the database
+const QUERY_CREATE = `
+    INSERT INTO project 
+        (sponsor_id, name, description) 
+    VALUES
+        ($1, $2, $3)
+    RETURNING id 
+`;
+
+// A query to replace an existing project. 
+const QUERY_UPDATE = `
+    UPDATE project 
+    SET 
+        (sponsor_id, name, description) = 
+        ($1, $2, $3)
+    WHERE 
+        id = $4
+    RETURNING id 
+`;
+
+/**
+ * Creates a Project row in the database
+ * 
+ * @param {*} client - node postgres client
+ * @param {*} sponsorId - the ID of the sponsor of the Project
+ * @param {*} name - The Name of the Project
+ * @param {*} description - The Project's Description
+ * 
+ * @returns {*} the ID of the created Project
+ */
+const createProject = async (client, sponsorId, name, description) => {
+    const values = [sponsorId, name, description];
+    return client.query(QUERY_CREATE, values);
+};
+
+/**
+ * Updates an existing project. This does a full replace of the old project.
+ * 
+ * @param {*} client - node postgres client
+ * @param {*} projectId - The ID of the Project to Update
+ * @param {*} sponsorId - the ID of the sponsor of the Project
+ * @param {*} name - The Name of the Project
+ * @param {*} description - The Project's Description
+ * 
+ * @returns {*} the ID of the updated Project
+ */
+const updateProject = async (client, projectId, sponsorId, name, description) => {
+    const values = [sponsorId, name, description, projectId];
+    return client.query(QUERY_UPDATE, values);
+};
+
+/**
+ * Finds projects that match the provided parameters. Gets all projects if all parameters are undefined.
+ * 
+ * @param {number} id the ID of the Project
+ * @param {number} sponsorId the ID of the Sponsor for the Project
+ * @param {string} name the name of the Project
+ * @returns {object[]|undefined} array of projects with fields (id, sponsor_id, name, and description), or undefined if all params are invalid.
+ * @throws error if the DB query failed to execute
+ * 
+ * @returns {*} an array of projects that match the parameters.
+ */
+const find = async (id, sponsorId, name) => {
+    let query = QUERY_FIND;
+    let values = [];
+    if ((typeof id !== 'undefined') && (id > 0)) {
+        values.push(id);
+        query = query + `AND id = $${values.length}` + ' ';
```

Comment:
I'm confused as to why you are using `values.length` instead of the `id` # for lines 90, 94, and 99?

Reply:
So this is building a prepared statement using the index of the parameter. See the "$1", "$2", etc from the other queries. You send the query string along with an array of values to the pg server where the pg server will do the insertion. 

In this case, if an id is not undefined, id will be pushed back onto the parameter list. The id parameter is now at index 1 (for some reason these start at 1 and not 0) which also happens to be the current length of the array. The server will substitute param at index 1 for "$1"